File: src/adapters/voyageai.ts
/**
 * Voyage AI embedder adapter (high-quality, cheaper than OpenAI).
 * Requires: VOYAGE_API_KEY
 *
 * Usage:
 *   import { AgentMemory } from 'agentic-memory';
 *   import { VoyageEmbedder } from 'agentic-memory/adapters/voyageai';
 *
 *   const memory = new AgentMemory({
 *     embedder: new VoyageEmbedder({ apiKey: process.env.VOYAGE_API_KEY }),
 *   });
 */

import type { Embedder } from '../core/types.js';

export interface VoyageEmbedderConfig {
  apiKey: string;
  model?: string;
}

export class VoyageEmbedder implements Embedder {
  private apiKey: string;
  private model: string;

  constructor(config: VoyageEmbedderConfig) {
    this.apiKey = config.apiKey;
    this.model = config.model ?? 'voyage-3-lite';
  }

  dimensions(): number {
    return this.model.includes('lite') ? 512 : 1024;
  }

  async embed(text: string): Promise<number[]> {
    const result = await this._call([text]);
    return result[0];
  }

  async embedBatch(texts: string[]): Promise<number[][]> {
    if (texts.length === 0) return [];
    return this._call(texts);
  }

  private async _call(input: string[]): Promise<number[][]> {
    const res = await fetch('https://api.voyageai.com/v1/embeddings', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${this.apiKey}`,
      },
      body: JSON.stringify({ model: this.model, input }),
    });

    if (!res.ok) {
      const err = await res.text();
      throw new Error(`Voyage AI embedding failed (${res.status}): ${err}`);
    }

    const data = await res.json() as { data: { embedding: number[] }[] };
    return data.data.map(d => d.embedding);
  }
}

File: python/agentic_memory/adapters/voyageai.py
"""
Voyage AI embedder adapter (high-quality, cheaper than OpenAI).
Requires: VOYAGE_API_KEY

Usage:
    from agentic_memory import AgentMemory
    from agentic_memory.adapters import VoyageEmbedder

    memory = AgentMemory(
        embedder=VoyageEmbedder(api_key="voyage-...")
    )
"""

from __future__ import annotations

import json
from urllib.request import Request, urlopen


class VoyageEmbedder:
    def __init__(
        self,
        api_key: str,
        model: str = "voyage-3-lite",
    ) -> None:
        self._api_key = api_key
        self._model = model

    def dimensions(self) -> int:
        return 512 if "lite" in self._model else 1024

    async def embed(self, text: str) -> list[float]:
        result = self._call([text])
        return result[0]

    async def embed_batch(self, texts: list[str]) -> list[list[float]]:
        if not texts:
            return []
        return self._call(texts)

    def _call(self, input_texts: list[str]) -> list[list[float]]:
        payload = json.dumps({
            "model": self._model,
            "input": input_texts,
        }).encode("utf-8")

        req = Request(
            "https://api.voyageai.com/v1/embeddings",
            data=payload,
            headers={
                "Content-Type": "application/json",
                "Authorization": f"Bearer {self._api_key}",
            },
            method="POST",
        )

        with urlopen(req) as resp:
            if resp.status != 200:
                raise RuntimeError(f"Voyage AI embedding failed ({resp.status})")
            data = json.loads(resp.read())

        return [d["embedding"] for d in data["data"]]
